File: LeetCode/Python/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#
# TODO: Solve (Binary Search)
# @lc code=start
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        l, r = 0, len(nums)
        lst = [-1, -1]
       
        while l + 1 < r:
            m = r + (l - r) // 2

            if nums[m] == target and nums[max(m - 1, 0)] != target:
                lst[0] = m
                l = m

            elif nums[m] == target and nums[min(m + 1, len(nums) - 1)] != target:
                lst[1] = m
                r = m

            elif nums[m] < target:
                r = m
            elif nums[m] > target:
                l = m

        if nums[l] == target:
            lst[0] = lst[1] = l
        if nums[r] == target:
            lst[0] = lst[1] = r
        
        print(lst)
        return lst
# @lc code=enda

File: LeetCode/CPP/34.find-first-and-last-position-of-element-in-sorted-array.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem34.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */
#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution
{
public:
    vector<int> searchRange(vector<int> &nums, int target)
    {
        vector<int> ans(2, -1);
        vector<int> first_check = binary_search(nums, target, 0, nums.size() - 1);

        if (first_check[0] == -1)
            return ans;

        vector<int> first_val = first_check, last_val = first_check;
        while (first_val[0] != -1)
        {
            ans[0] = first_val[1];
            first_val = binary_search(nums, target, first_val[0], first_val[1] - 1);
        }

        while (last_val[0] != -1)
        {
            ans[1] = last_val[1];
            last_val = binary_search(nums, target, last_val[1] + 1, last_val[2]);
        }

        return ans;
    }

    vector<int> binary_search(vector<int> &nums, int target, int low, int high)
    {
        int m;
        while (low <= high)
        {
            m = low + (high - low) / 2;
            if (nums[m] == target)
            {
                return vector<int>{low, m, high};
            }
            else if (nums[m] > target)
            {
                high = m - 1;
            }
            else
            {
                low = m + 1;
            }
        }
        return vector<int>{-1, -1, -1};
    }
};

// @lc code=end